src/refda/exec.c

```diff
@@ -452,3 +452,171 @@ void *refda_exec_worker(void *arg)
+        //  and schedule the next call there, once execution is truly finished
+        //
+        //  That completely avoids a (unlikely?) situation where execution is still ongoing
+        //  when the next iteration is scheduled
+        //
+        // Anyway, the call we want to make is this, to schedule another call using "period" to compute the time offset:
+        //
+        // refda_exec_schedule_tbr(agent, tbr, false)
+    }
+
+    return exec_result;
+}
+
+/** Compute the next scheduled time at which to run the TBR
+ */
+static int refda_exec_tbr_next_scheduled_time(struct timespec *schedtime, const refda_amm_tbr_desc_t *tbr,
+                                              bool starting)
+{
+    if (starting)
+    {
+        if (!tbr->start_time.is_ref && tbr->start_time.as_lit.ari_type == CACE_ARI_TYPE_TP)
+        {
+            *schedtime = tbr->start_time.as_lit.value.as_timespec;
+        }
+        else if (!tbr->start_time.is_ref && tbr->start_time.as_lit.ari_type == CACE_ARI_TYPE_TP)
+        {
+            struct timespec reltime = tbr->start_time.as_lit.value.as_timespec;
+            if (reltime.tv_nsec == 0 && reltime.tv_sec == 0)
+            {
+                // Rule is always active, start it now
+                clock_gettime(CLOCK_REALTIME, schedtime);
+            }
+            else
+            {
+                // TODO:
+                // A relative start time SHALL be interpreted relative to the absolute time at which the
+                // Agent is initialized (for ADM rules) or the rule is created (for ODM rules).
+            }
+        }
+        else
+        {
+            CACE_LOG_ERR("Invalid start time for TBR");
+            return 2;
+        }
+    }
+    else
+    {
+        clock_gettime(CLOCK_REALTIME, schedtime);
+        *schedtime = timespec_add(*schedtime, tbr->period.as_lit.value.as_timespec);
+    }
+
+    return 0;
+}
+
+/**
+ * Schedule execution of a time based rule
+ */
+static int refda_exec_schedule_tbr(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr, bool starting)
+{
+    refda_exec_item_t *item = NULL; // TODO
+
+    struct timespec schedtime;
+    int             result = refda_exec_tbr_next_scheduled_time(&schedtime, tbr, starting);
+    if (!result)
+    {
+        /**
+         * TODO: need to do a few things here
+         * - create a new type of item (or modify existing) such that it can hold either a CTRL or a rule
+         * - then, we pass an instance of that item as part of the event.agent
+         *   - TBD: who owns data pointed to by item?? Maybe it is part of TBR and is held there
+         * - need to support a different type signature for callback. could expand existing
+         *   signature to support either a CTRL or a rule
+         */
+        refda_timeline_event_t event = { .ts   = schedtime,
+                                         .item = (refda_exec_item_t *)tbr, // TODO: this doesn't work, need a container
```

Comment:
Feel free to add an enum field to the timeline event for a "purpose" or similar, with a union of per-purpose data struct. There are examples of this logic in `cace_ari_idseg_t` and some others where C `union` is used in CACE.

Reply:
Added a `purpose` field and moved CTRL / TBR specific fields to structures behind a `union`.